cachix/src/Cachix/Client/Store.hs

```diff
@@ -0,0 +1,127 @@
+
+data PathInfo = PathInfo
+  { deriver :: Maybe Text,
+    narSize :: Int64,
+    narHash :: Text,
+    references :: [Text]
+  }
+
+followLinksToStorePath :: Store -> FilePath -> IO FilePath
+followLinksToStorePath (Store prefix _) path = do
+  storePath <- canonicalizePath path
+  let storePath' = T.drop (T.length prefix) (toS storePath)
+  return $ toS $ prefix <> T.intercalate "/" (take 3 $ T.splitOn "/" storePath')
+
+withStore :: Text -> (Store -> IO ()) -> IO ()
+withStore storePrefix =
+  bracket open close
+  where
+    uri = "file:" <> toS storePrefix <> "/var/nix/db/db.sqlite?immutable=1"
+    flags = [SQLite.SQLOpenReadOnly, SQLite.SQLOpenURI]
+    close (Store _ db) = SQLite.close db
+    open = do
+      (_, out, _) <- readProcessWithExitCode "nix" ["show-config", "--extra-experimental-features", "nix-command"] mempty
+      let vfs =
+            if "use-sqlite-wal = false" `T.isInfixOf` toS out
+              then SQLite.SQLVFSUnixDotFile
+              else SQLite.SQLVFSDefault
+      conn <- SQLite.open2 uri flags vfs
+      return $ Store storePrefix conn
+
+queryNarinfo :: Text
+queryNarinfo = "select id, hash, deriver, narSize from ValidPaths where path = :path"
+
+queryReferences :: Text
+queryReferences = "select path from Refs join ValidPaths on reference = id where referrer = :id"
+
+query :: Store -> Text -> [(Text, SQLData)] -> IO [[SQLite.SQLData]]
+query (Store _ conn) txt bindings = do
+  stmt <- SQLite.prepare conn txt
+  SQLite.bindNamed stmt bindings
+  getRows stmt
+
+getRows :: SQLite.Statement -> IO [[SQLite.SQLData]]
+getRows stmt = do
+  SQLite.step stmt >>= \case
+    SQLite.Row -> do
+      row <- SQLite.columns stmt
+      rows <- getRows stmt
+      return $ row : rows
+    SQLite.Done -> do
+      SQLite.finalize stmt
+      return []
+
+queryPathInfo :: Store -> Text -> IO (Either Text PathInfo)
+queryPathInfo store path = do
+  rows <- query store queryNarinfo [(":path", SQLite.SQLText path)]
+  case rows of
+    [] -> return $ Left $ "no such path " <> path
+    [[id_, SQLite.SQLText hash_, deriver, SQLite.SQLInteger narSize]] -> do
+      references <- query store queryReferences [(":id", id_)]
+      refs <- traverse go references
+      return $
+        Right $
+          PathInfo
+            { deriver = getDeriver deriver,
+              narSize = narSize,
+              narHash = hash_,
+              references = refs
+            }
+    _ -> return $ Left $ "got invalid narinfo from nix " <> show rows
+  where
+    go [SQLite.SQLText path_] = return path_
+    go a = throwIO $ FatalError $ "invalid reference type " <> show a
+
+    getDeriver :: SQLite.SQLData -> Maybe Text
+    getDeriver (SQLite.SQLText deriver) = Just deriver
+    getDeriver _ = Nothing
+
+computeClosure :: Store -> [StorePath] -> IO [StorePath]
+computeClosure store initialPaths = do
+  allPaths <- processGraph (getPath <$> initialPaths) (fmap (concatMap references . rights . List.singleton) . queryPathInfo store)
```

Comment:
Do we care about any of the errors returned by `queryPathInfo` here? If not, should we document the reason?

What if we avoid going through a list here? The `.` could also be a `<$>` to be read as the common double fmap pattern.
```suggestion
  let fetchReferences = fmap (either (const []) references) . queryPathInfo store
  allPaths <- processGraph (getPath <$> initialPaths) fetchReferences
```

Or maybe even something really obvious:
```suggestion
  allPaths <- 
    processGraph (getPath <$> initialPaths) $
      queryPathInfo store >=> \case
        Left _ -> pure []
        Right pathInfo -> pure $ references pathInfo
```